lib/src/parsers/phone_parser.dart

```diff
@@ -102,15 +104,50 @@ abstract class PhoneParser {
         phoneWithoutExitCode.startsWith(callerNationalPrefix)) {
       return callerMetadata;
     }
-    // if no caller was provided we need to make a best guess given the country code
-    final countryCode =
-        CountryCodeParser.extractCountryCode(phoneWithoutExitCode);
-    final national =
-        CountryCodeParser.removeCountryCode(phoneWithoutExitCode, countryCode);
-    // multiple countries use the same country code
-    final metadatas = MetadataFinder.getMetadatasForCountryCode(countryCode);
-    // if multiple countries share the same country code, patterns on the national number
-    // is used to find the correct country.
-    return MetadataMatcher.getMatchUsingPatterns(national, metadatas);
+
+    String? countryCode;
+    Object? countryCodeException;
+    List<PhoneMetadata>? metadatas;
+
+    try {
+      countryCode = CountryCodeParser.extractCountryCode(phoneWithoutExitCode);
+    } catch(e) {
+      countryCodeException = e;
+    }
+
+    if (countryCode != null) {
+      metadatas = MetadataFinder.getMetadatasForCountryCode(countryCode);
+      final national = CountryCodeParser.removeCountryCode(phoneWithoutExitCode, countryCode);
+      final metadataMatch = MetadataMatcher.getMatchUsingPatternsStrict(national, metadatas);
+
+      if (phoneHadExitCode) {
+        // 1.
+        // if phone number has the exit code, we should use it 
+        // despite caller's metadata and phone number validity
+        return metadataMatch ?? metadatas.first;
+      }
+
+      // 2.
+      // if number has no exit code. Try to figure out if it
+      // starts with a country code
+      if (metadataMatch != null) {
+        return metadataMatch;
+      }
+    }
+
+    // 3.
+    // If any direct match was't found, use provided callerMetadata
+    if (callerMetadata != null) {
+      return callerMetadata;
+    }
+
+    if (metadatas != null) {
+      // 4.
+      // The last resort. Use any of found metadatas ignoring validity
+      return metadatas.first;
+    }
+
+    // rethrow previously got error
+    throw countryCodeException!;

```

Comment:
let's not use the !

Reply:
That is not possible because it leads to compilation issues. This can be avoided with the transformation `CountryCodeParser.extractCountryCode` and `_findDestinationMetadata` to return an optional, which leads to more remarkable code changes.

Reply:
It seems like you were talking about the country_code_parser interface that throws when the country code is not found. Which could instead return a `String?`. That's a fair observation and I'm okay with that. At first glance this seems to be a small change but it'd be better if it was in a separate PR imo, just to make it easy to review.

Reply:
Agreed. What do you think is the best to do in this case?